packages/contracts/contracts/testing/ERC20Reward.sol

```diff
@@ -21,4 +22,46 @@ contract ERC20Reward is ERC20, ERC20Burnable, ERC20Permit, ERC7529 {
     function mint(address to, uint256 amount) public onlyRole(MINTER_ROLE) {
         _mint(to, amount);
     }
+
+    function redeem(address account, uint256 value, bytes memory signature) public {
+
+        // Regenerate the message 
+        bytes32 hash = ECDSA.toEthSignedMessageHash(
+            keccak256(abi.encodePacked(address(this), account, value))
+        );
+
+        // Get the signer of the message
+        address signer = _getSigner(hash, signature);
+
+        // If the caller has is an admin, check the signer against the target account to burn tokens from
+        // If the caller is not an admin, check that the function caller is the signer
+        if (hasRole(DEFAULT_ADMIN_ROLE, _msgSender())) {
+            require(signer == account, "ERC20Reward: Signer does not match account to redeem from");
+        } else {
+            require(signer == _msgSender(), "ERC20Reward: Caller did not sign the message"); 
```

Comment:
I am having hard times digesting this, I thought that signer == account should always be there wether the caller is admin or not and then if the caller if not an admin  _msgSender() == account is also needed as only admins can call this message to burn other accounts tokens, I may be getting this wrong though so let's discuss this off chat when you have some time @seansing9

Reply:
The scenario running in my head is like the following:

1. Account A (the attacker) generates a signed message that authorizes a token burn from Account B (the victim). This message includes the contract address, Account B's address as the account from which tokens should be redeemed, and the amount of tokens to burn.

2. Account A then calls the redeem function with this signed message.

if the Account A is not an admin, signer == _msgSender() will get him passed as he's the account who signed the signature and called the redeem method and therefore burns tokens for Account B without being authorized.

Please let me know if I am missing something here